File: surface/widgets/winui/components/widgets/concepts/llm-agent/AgentTimeline.cs
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Input;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI;

namespace Clef.Surface.Widgets.Concepts.LlmAgent
{
    public sealed partial class AgentTimeline : UserControl
    {
        private enum WidgetState { Idle, EntrySelected, Interrupted, Inactive, Active }
        private enum WidgetEvent { NewEntry, SelectEntry, Interrupt, Deselect, Resume, StreamStart, StreamEnd }

        private WidgetState _state = WidgetState.Idle;
        private int _selectedIndex = -1;
        private string _typeFilter = null;
        private readonly StackPanel _root;
        private readonly StackPanel _headerPanel;
        private readonly TextBlock _agentBadge;
        private readonly TextBlock _statusIndicator;
        private readonly Button _interruptButton;
        private readonly StackPanel _filterBar;
        private readonly TextBlock _interruptBanner;
        private readonly StackPanel _timelinePanel;
        private readonly ScrollViewer _scrollViewer;

        private static readonly string[] AllTypes = { "thought", "tool-call", "tool-result", "response", "error" };
        private static readonly string[] TypeIcons = { "\u2022\u2022\u2022", "\u2699", "\u2611", "\u25B6", "\u2717" };
        private static readonly string[] TypeLabels = { "Thought", "Tool Call", "Tool Result", "Response", "Error" };

        public AgentTimeline()
        {
            _root = new StackPanel { Orientation = Orientation.Vertical, Spacing = 8 };

            // Header
            _headerPanel = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8 };
            _agentBadge = new TextBlock { Text = "Agent", FontWeight = Microsoft.UI.Text.FontWeights.Bold, FontSize = 14 };
            _statusIndicator = new TextBlock { Text = "\u25CB idle", FontSize = 13 };
            _interruptButton = new Button { Content = "Interrupt", Visibility = Visibility.Collapsed };
            _interruptButton.Click += (s, e) => Send(WidgetEvent.Interrupt);
            _headerPanel.Children.Add(_agentBadge);
            _headerPanel.Children.Add(_statusIndicator);
            _headerPanel.Children.Add(_interruptButton);
            _root.Children.Add(_headerPanel);

            // Filter bar
            _filterBar = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 4 };
            var allBtn = new Button { Content = "All" };
            allBtn.Click += (s, e) => { _typeFilter = null; };
            _filterBar.Children.Add(allBtn);
            for (int i = 0; i < AllTypes.Length; i++)
            {
                var idx = i;
                var btn = new Button { Content = $"{TypeIcons[idx]} {TypeLabels[idx]}" };
                btn.Click += (s, e) => { _typeFilter = _typeFilter == AllTypes[idx] ? null : AllTypes[idx]; };
                _filterBar.Children.Add(btn);
            }
            _root.Children.Add(_filterBar);

            // Interrupt banner
            _interruptBanner = new TextBlock
            {
                Text = "Agent execution interrupted",
                Visibility = Visibility.Collapsed,
                Foreground = new SolidColorBrush(Colors.Red),
                FontSize = 13
            };
            AutomationProperties.SetLiveSetting(_interruptBanner, AutomationLiveSetting.Assertive);
            _root.Children.Add(_interruptBanner);

            // Timeline entries
            _timelinePanel = new StackPanel { Orientation = Orientation.Vertical, Spacing = 4 };
            _scrollViewer = new ScrollViewer { Content = _timelinePanel, MaxHeight = 400 };
            _root.Children.Add(_scrollViewer);

            this.Content = _root;
            AutomationProperties.SetName(this, "Multi-agent communication timeline display");
        }

        public void SetAgent(string agentName, string status)
        {
            _agentBadge.Text = agentName;
            _statusIndicator.Text = status == "running" ? "\u25CF running" : $"\u25CB {status}";
            _interruptButton.Visibility = status == "running" ? Visibility.Visible : Visibility.Collapsed;
        }

        public void AddEntry(string type, string label, string timestamp, string duration = null, string status = "complete")
        {
            int typeIdx = System.Array.IndexOf(AllTypes, type);
            string icon = typeIdx >= 0 ? TypeIcons[typeIdx] : "\u2022";

            var entryPanel = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8, Padding = new Thickness(4) };
            var iconText = new TextBlock { Text = icon, FontSize = 13, VerticalAlignment = VerticalAlignment.Center };
            var labelText = new TextBlock { Text = label, FontSize = 13, VerticalAlignment = VerticalAlignment.Center };
            var timeText = new TextBlock { Text = timestamp, FontSize = 11, Opacity = 0.7, VerticalAlignment = VerticalAlignment.Center };

            entryPanel.Children.Add(iconText);
            entryPanel.Children.Add(labelText);

            if (status == "running")
            {
                var runningIndicator = new ProgressRing { IsActive = true, Width = 14, Height = 14 };
                entryPanel.Children.Add(runningIndicator);
            }

            if (duration != null && status != "running")
            {
                entryPanel.Children.Add(new TextBlock { Text = duration, FontSize = 11, Opacity = 0.6 });
            }

            entryPanel.Children.Add(timeText);

            string typeLabel = typeIdx >= 0 ? TypeLabels[typeIdx] : type;
            AutomationProperties.SetName(entryPanel, $"{typeLabel}: {label}");
            _timelinePanel.Children.Add(entryPanel);

            Send(WidgetEvent.NewEntry);
        }

        public void Send(WidgetEvent evt)
        {
            _state = Reduce(_state, evt);
            _interruptBanner.Visibility = _state == WidgetState.Interrupted ? Visibility.Visible : Visibility.Collapsed;
        }

        private WidgetState Reduce(WidgetState state, WidgetEvent evt) => state switch
        {
            WidgetState.Idle when evt == WidgetEvent.NewEntry => WidgetState.Idle,
            WidgetState.Idle when evt == WidgetEvent.SelectEntry => WidgetState.EntrySelected,
            WidgetState.Idle when evt == WidgetEvent.Interrupt => WidgetState.Interrupted,
            WidgetState.EntrySelected when evt == WidgetEvent.Deselect => WidgetState.Idle,
            WidgetState.EntrySelected when evt == WidgetEvent.SelectEntry => WidgetState.EntrySelected,
            WidgetState.Interrupted when evt == WidgetEvent.Resume => WidgetState.Idle,
            WidgetState.Inactive when evt == WidgetEvent.StreamStart => WidgetState.Active,
            WidgetState.Active when evt == WidgetEvent.StreamEnd => WidgetState.Inactive,
            _ => state
        };
    }
}

File: surface/widgets/nativescript/components/widgets/concepts/llm-agent/AgentTimeline.ts
import { StackLayout, Label, Button, ScrollView, FlexboxLayout, ActivityIndicator } from '@nativescript/core';

/* ---------------------------------------------------------------------------
 * AgentTimeline state machine
 * ------------------------------------------------------------------------- */

export type AgentTimelineState = 'idle' | 'entrySelected' | 'interrupted' | 'inactive' | 'active';
export type AgentTimelineEvent =
  | { type: 'NEW_ENTRY' }
  | { type: 'SELECT_ENTRY'; id?: string }
  | { type: 'INTERRUPT' }
  | { type: 'DESELECT' }
  | { type: 'RESUME' }
  | { type: 'STREAM_START' }
  | { type: 'STREAM_END' };

export function agentTimelineReducer(state: AgentTimelineState, event: AgentTimelineEvent): AgentTimelineState {
  switch (state) {
    case 'idle':
      if (event.type === 'NEW_ENTRY') return 'idle';
      if (event.type === 'SELECT_ENTRY') return 'entrySelected';
      if (event.type === 'INTERRUPT') return 'interrupted';
      return state;
    case 'entrySelected':
      if (event.type === 'DESELECT') return 'idle';
      if (event.type === 'SELECT_ENTRY') return 'entrySelected';
      return state;
    case 'interrupted':
      if (event.type === 'RESUME') return 'idle';
      return state;
    case 'inactive':
      if (event.type === 'STREAM_START') return 'active';
      return state;
    case 'active':
      if (event.type === 'STREAM_END') return 'inactive';
      return state;
    default:
      return state;
  }
}

/* ---------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

export type EntryType = 'thought' | 'tool-call' | 'tool-result' | 'response' | 'error';
export type EntryStatus = 'running' | 'complete' | 'error';

export interface TimelineEntry {
  id: string;
  type: EntryType;
  label: string;
  timestamp: string;
  duration?: number;
  detail?: string;
  status?: EntryStatus;
}

export interface AgentTimelineProps {
  entries: TimelineEntry[];
  agentName: string;
  status: string;
  showDelegations?: boolean;
  autoScroll?: boolean;
  maxEntries?: number;
  onInterrupt?: () => void;
}

/* ---------------------------------------------------------------------------
 * Helpers
 * ------------------------------------------------------------------------- */

const TYPE_ICONS: Record<EntryType, string> = {
  'thought': '\u2022\u2022\u2022',
  'tool-call': '\u2699',
  'tool-result': '\u2611',
  'response': '\u25B6',
  'error': '\u2717',
};

const TYPE_LABELS: Record<EntryType, string> = {
  'thought': 'Thought',
  'tool-call': 'Tool Call',
  'tool-result': 'Tool Result',
  'response': 'Response',
  'error': 'Error',
};

function formatDuration(ms: number): string {
  if (ms < 1000) return `${ms}ms`;
  return `${(ms / 1000).toFixed(1)}s`;
}

/* ---------------------------------------------------------------------------
 * Widget
 * ------------------------------------------------------------------------- */

export function createAgentTimeline(props: AgentTimelineProps): { view: StackLayout; dispose: () => void } {
  const {
    entries,
    agentName,
    status,
    maxEntries = 100,
    onInterrupt,
  } = props;

  let widgetState: AgentTimelineState = 'idle';
  let selectedEntryId: string | null = null;
  let expandedIds = new Set<string>();
  let typeFilter: EntryType | null = null;
  const disposers: (() => void)[] = [];

  function send(event: AgentTimelineEvent) {
    widgetState = agentTimelineReducer(widgetState, event);
    update();
  }

  const root = new StackLayout();
  root.className = 'agent-timeline';
  root.automationText = `Agent timeline: ${agentName}`;

  // Header
  const header = new FlexboxLayout();
  header.className = 'agent-timeline-header';
  header.flexDirection = 'row' as any;
  header.alignItems = 'center' as any;

  const agentBadge = new Label();
  agentBadge.className = 'agent-timeline-agent-badge';
  agentBadge.text = agentName;
  header.addChild(agentBadge);

  const statusLabel = new Label();
  statusLabel.className = 'agent-timeline-status';
  statusLabel.text = `${status === 'running' ? '\u25CF' : '\u25CB'} ${status}`;
  header.addChild(statusLabel);

  const interruptBtn = new Button();
  interruptBtn.className = 'agent-timeline-interrupt';
  interruptBtn.text = 'Interrupt';
  interruptBtn.automationText = 'Interrupt agent';
  interruptBtn.visibility = (status === 'running' ? 'visible' : 'collapse') as any;
  const interruptHandler = () => {
    send({ type: 'INTERRUPT' });
    onInterrupt?.();
  };
  interruptBtn.on('tap', interruptHandler);
  disposers.push(() => interruptBtn.off('tap', interruptHandler));
  header.addChild(interruptBtn);

  root.addChild(header);

  // Filter bar
  const filterBar = new FlexboxLayout();
  filterBar.className = 'agent-timeline-filter-bar';
  filterBar.flexDirection = 'row' as any;

  const allFilterBtn = new Button();
  allFilterBtn.className = 'agent-timeline-filter-button';
  allFilterBtn.text = 'All';
  const allFilterHandler = () => { typeFilter = null; rebuildEntries(); };
  allFilterBtn.on('tap', allFilterHandler);
  disposers.push(() => allFilterBtn.off('tap', allFilterHandler));
  filterBar.addChild(allFilterBtn);

  const allTypes: EntryType[] = ['thought', 'tool-call', 'tool-result', 'response', 'error'];
  for (const t of allTypes) {
    const btn = new Button();
    btn.className = 'agent-timeline-filter-button';
    btn.text = `${TYPE_ICONS[t]} ${TYPE_LABELS[t]}`;
    const handler = () => {
      typeFilter = typeFilter === t ? null : t;
      rebuildEntries();
    };
    btn.on('tap', handler);
    disposers.push(() => btn.off('tap', handler));
    filterBar.addChild(btn);
  }

  root.addChild(filterBar);

  // Interrupted banner
  const interruptBanner = new Label();
  interruptBanner.className = 'agent-timeline-interrupt-banner';
  interruptBanner.text = 'Agent execution interrupted';
  interruptBanner.visibility = 'collapse' as any;
  root.addChild(interruptBanner);

  // Timeline scroll area
  const scrollView = new ScrollView();
  scrollView.className = 'agent-timeline-scroll';

  const timelineContainer = new StackLayout();
  timelineContainer.className = 'agent-timeline-entries';
  scrollView.content = timelineContainer;
  root.addChild(scrollView);

  function rebuildEntries() {
    timelineContainer.removeChildren();
    const limited = entries.slice(-maxEntries);
    const visible = typeFilter ? limited.filter((e) => e.type === typeFilter) : limited;

    for (const entry of visible) {
      const entryView = new StackLayout();
      entryView.className = 'agent-timeline-entry';
      entryView.automationText = `${TYPE_LABELS[entry.type]}: ${entry.label}`;

      const entryRow = new FlexboxLayout();
      entryRow.flexDirection = 'row' as any;
      entryRow.alignItems = 'center' as any;

      // Type badge
      const badge = new Label();
      badge.className = 'agent-timeline-type-badge';
      badge.text = TYPE_ICONS[entry.type];
      entryRow.addChild(badge);

      // Label
      const entryLabel = new Label();
      entryLabel.className = 'agent-timeline-entry-label';
      entryLabel.text = entry.label;
      entryRow.addChild(entryLabel);

      // Running indicator
      if (entry.status === 'running') {
        const indicator = new ActivityIndicator();
        indicator.className = 'agent-timeline-running-indicator';
        indicator.busy = true;
        indicator.width = 16;
        indicator.height = 16;
        entryRow.addChild(indicator);
      }

      // Duration
      if (entry.duration != null && entry.status !== 'running') {
        const durationLabel = new Label();
        durationLabel.className = 'agent-timeline-duration';
        durationLabel.text = formatDuration(entry.duration);
        entryRow.addChild(durationLabel);
      }

      // Timestamp
      const tsLabel = new Label();
      tsLabel.className = 'agent-timeline-timestamp';
      tsLabel.text = entry.timestamp;
      entryRow.addChild(tsLabel);

      entryView.addChild(entryRow);

      // Expanded detail
      if (expandedIds.has(entry.id) && entry.detail) {
        const detailLabel = new Label();
        detailLabel.className = 'agent-timeline-detail';
        detailLabel.text = entry.detail;
        detailLabel.textWrap = true;
        entryView.addChild(detailLabel);
      }

      // Tap to select/expand
      const tapHandler = () => {
        selectedEntryId = entry.id;
        if (expandedIds.has(entry.id)) {
          expandedIds.delete(entry.id);
        } else {
          expandedIds.add(entry.id);
        }
        send({ type: 'SELECT_ENTRY', id: entry.id });
        rebuildEntries();
      };
      entryView.on('tap', tapHandler);

      timelineContainer.addChild(entryView);
    }
  }

  rebuildEntries();

  function update() {
    interruptBanner.visibility = (widgetState === 'interrupted' ? 'visible' : 'collapse') as any;
  }

  return {
    view: root,
    dispose() {
      disposers.forEach((d) => d());
    },
  };
}

export default createAgentTimeline;
